File: kmg-web/src/types/project.ts
import type { StockMovementDTO } from "./stock";
import type { CheckStatus } from "./supplier";

export interface ProjectListDTO {
  id: number;
  name: string;
  projectCode: string;
  projectType: string;
  status: string;
  clientName: string;
  contractValue: number;
  totalCollected: number;
  remainingBalance: number;
  netProfit: number;
  createdAt: string;
}

export interface ProjectPaymentDTO {
  id: number;
  amount: number;
  amountCash: number;
  amountCredit: number;
  paymentDate: string;
  notes: string | null;
  attachmentUrl: string | null;
  attachmentFileName: string | null;
  isCheck: boolean;
  checkDueDate: string | null;
  checkStatus: CheckStatus | null;
}

export interface ProjectExpenseDTO {
  id: number;
  amount: number;
  category: string;
  description: string | null;
  expenseDate: string;
  attachmentUrl: string | null;
  attachmentFileName: string | null;
}

export interface ProjectAttachmentDTO {
  id: number;
  fileUrl: string;
  fileName: string | null;
  description: string | null;
  uploadedAt: string;
  uploadedByEmployeeName: string;
}

export interface ProjectAuditDTO {
  id: number;
  actionDescription: string;
  actionDate: string;
  employeeName: string;
}

export interface ProjectDetailsDTO extends ProjectListDTO {
  clientId: number;
  clientPhone: string | null;
  clientEmail: string | null;
  clientAddress: string | null;
  description: string | null;
  tenderInsurancePercent: number | null;
  tenderInsuranceAmount: number;
  insuranceDueDate: string | null;
  insuranceRecovered: boolean;
  tenderTaxPercent: number | null;
  tenderTaxAmount: number;
  workGuaranteePercent: number | null;
  workGuaranteeAmount: number;
  workGuaranteeDueDate: string | null;
  workGuaranteeRecovered: boolean;
  contractValueWithTax: number;
  supplyProfitMargin: number | null;
  totalMaterialsCost: number;
  totalPettyExpenses: number;
  totalLaborCost: number;
  totalWriteOffs: number;
  payments: ProjectPaymentDTO[];
  expenses: ProjectExpenseDTO[];
  stockMovements: StockMovementDTO[];
  attachments: ProjectAttachmentDTO[];
  auditLogs: ProjectAuditDTO[];
  writeOffs: ProjectWriteOffDTO[];
  canDelete: boolean;
}

export interface ProjectWriteOffDTO {
  id: number;
  amount: number;
  reason: string;
  writeOffDate: string;
}

export interface CreateProjectWriteOffDTO {
  projectId: number;
  amount: number;
  reason: string;
  writeOffDate: string;
}

export interface UpdateProjectWriteOffDTO {
  id: number;
  amount: number;
  reason: string;
  writeOffDate: string;
}

export interface CreateProjectDTO {
  name: string;
  projectType: number;
  clientId: number;
  contractValue: number;
  description?: string | null;
  tenderInsurancePercent?: number | null;
  insuranceDueDate?: string | null;
  tenderTaxPercent?: number | null;
  workGuaranteePercent?: number | null;
  workGuaranteeDueDate?: string | null;
  supplyProfitMargin?: number | null;
}

export interface UpdateProjectDTO {
  id: number;
  name: string;
  description?: string | null;
  clientId: number;
  contractValue: number;
  tenderInsurancePercent?: number | null;
  insuranceDueDate?: string | null;
  tenderTaxPercent?: number | null;
  workGuaranteePercent?: number | null;
  workGuaranteeDueDate?: string | null;
}

export interface UpdateProjectStatusDTO {
  projectId: number;
  status: number;
}

export interface CreateProjectPaymentDTO {
  projectId: number;
  amountCash: number;
  amountCredit: number;
  paymentDate: string;
  notes?: string | null;
  attachmentUrl?: string | null;
  attachmentFileName?: string | null;
  isCheck: boolean;
  checkDueDate?: string | null;
}

export interface UpdateProjectPaymentDTO {
  id: number;
  amountCash: number;
  amountCredit: number;
  paymentDate: string;
  notes?: string | null;
  attachmentUrl?: string | null;
  attachmentFileName?: string | null;
  isCheck: boolean;
  checkDueDate?: string | null;
}

export interface ResolveProjectPaymentCheckDTO {
  paymentId: number;
  action: number;
  newDueDate?: string | null;
}

export interface CreateProjectExpenseDTO {
  projectId: number;
  amount: number;
  category: number;
  description?: string | null;
  expenseDate: string;
  attachmentUrl?: string | null;
  attachmentFileName?: string | null;
  paidFromCashBox: boolean;
}

export interface UpdateProjectExpenseDTO {
  id: number;
  amount: number;
  category: number;
  description?: string | null;
  expenseDate: string;
}

export interface CreateProjectAttachmentDTO {
  projectId: number;
  fileUrl: string;
  fileName?: string | null;
  description?: string | null;
}

export interface UpdateProjectAttachmentDTO {
  id: number;
  description?: string | null;
}

File: KMG.Core/Models/Project.cs
using System.ComponentModel.DataAnnotations.Schema;
using KMG.Core.Enums;

namespace KMG.Core.Models
{
    public class Project
    {
        public int Id { get; set; }
        public string Name { get; set; } = string.Empty;
        public string ProjectCode { get; set; } = string.Empty;
        public ProjectType ProjectType { get; set; }
        public ProjectStatus Status { get; set; }
        public string? Description { get; set; }

        public decimal ContractValue { get; set; }

        // خاص بمشاريع المناقصات (تصنيع وتنفيذ) - نسب مش قيم مطلقة، بتتحسب على قيمة العقد
        public decimal? TenderInsurancePercent { get; set; }
        public decimal? TenderTaxPercent { get; set; }
        public decimal? WorkGuaranteePercent { get; set; }

        public DateTime? InsuranceDueDate { get; set; }
        public bool InsuranceRecovered { get; set; }
        public DateTime? WorkGuaranteeDueDate { get; set; }
        public bool WorkGuaranteeRecovered { get; set; }

        // خاص بمشاريع التوريد
        public decimal? SupplyProfitMargin { get; set; }

        public DateTime CreatedAt { get; set; }

        public int ClientId { get; set; }
        public virtual Client Client { get; set; } = null!;

        public int CreatedByEmployeeId { get; set; }
        public virtual Employee CreatedByEmployee { get; set; } = null!;

        public virtual ICollection<ProjectPayment> Payments { get; set; } = new List<ProjectPayment>();
        public virtual ICollection<ProjectExpense> Expenses { get; set; } = new List<ProjectExpense>();
        public virtual ICollection<StockMovement> StockMovements { get; set; } = new List<StockMovement>();
        public virtual ICollection<Mission> Missions { get; set; } = new List<Mission>();
        public virtual ICollection<ProjectAttachment> Attachments { get; set; } = new List<ProjectAttachment>();
        public virtual ICollection<ProjectAudit> AuditLogs { get; set; } = new List<ProjectAudit>();
        public virtual ICollection<ProjectWriteOff> WriteOffs { get; set; } = new List<ProjectWriteOff>();
        public virtual ICollection<CashBoxTransaction> CashBoxTransactions { get; set; } = new List<CashBoxTransaction>();

        // تكلفة الخامات = (صرف للمشروع - مرتجع من المشروع) بسعر وقت الحركة
        [NotMapped]
        public decimal TotalMaterialsCost =>
            StockMovements.Where(m => m.MovementType == MovementType.IssueToProject).Sum(m => m.Quantity * m.UnitPriceAtTime)
            - StockMovements.Where(m => m.MovementType == MovementType.ReturnFromProject).Sum(m => m.Quantity * m.UnitPriceAtTime);

        [NotMapped]
        public decimal TotalPettyExpenses => Expenses.Sum(e => e.Amount);

        [NotMapped]
        public decimal TotalLaborCost => Missions.Sum(m => m.TotalLaborCost);

        [NotMapped]
        public decimal TotalCollected => Payments.Sum(p => p.Amount);

        [NotMapped]
        public decimal TotalWriteOffs => WriteOffs.Sum(w => w.Amount);

        [NotMapped]
        public decimal TenderInsuranceAmount => TenderInsurancePercent is > 0 ? ContractValue * TenderInsurancePercent.Value / 100m : 0;

        [NotMapped]
        public decimal TenderTaxAmount => TenderTaxPercent is > 0 ? ContractValue * TenderTaxPercent.Value / 100m : 0;

        [NotMapped]
        public decimal WorkGuaranteeAmount => WorkGuaranteePercent is > 0 ? ContractValue * WorkGuaranteePercent.Value / 100m : 0;

        // الضريبة بتزود المبلغ المطلوب تحصيله فعليًا من العميل - مش رقم عرض بس
        [NotMapped]
        public decimal ContractValueWithTax => ContractValue + TenderTaxAmount;

        [NotMapped]
        public decimal RemainingBalance => ContractValueWithTax - TotalCollected - TotalWriteOffs;

        [NotMapped]
        public decimal NetProfit => ContractValue - (TotalMaterialsCost + TotalPettyExpenses + TotalLaborCost);
    }
}
